### ISSUE TITLE
Infer generics in `is` check for type promotion

### ISSUE THREAD

**Issue by rkj** [dart-lang/sdk#34018]
Consider following code:

```dart
abstract class A<T> {}
abstract class B<T> implements A<T> {
  b();
}

void foo<T>(A<T> a) {
  if (a is B) {
    return a.b();
  }
}
```

It gives error message:

>  // The method 'b' isn't defined for the class 'A<T>'.

so it seems the type promotion on `a` didn't happen inside the `if` block. To fix this I can change the condition to be: `if (a is B<T>)`.

This somewhat makes sense, but it seems to me that in this case compiler could infer that `B` must be `B<T>`, if it's casted from `A<T>`. I guess it could be some subclass of `T`, but that would still be substitutable for B. Or maybe there are some other edge cases in general. If this is not possible then maybe compiler could give some hint why type promotion didn't happen?

**Comment by srawlins** [comment]
I  don't think this belongs to `core-l`; it's a language request, correct?

**Comment by tp** [comment]
I just ran into a similar issue:

```dart
abstract class I {
    int get value;
}

class X extends Route<dynamic> implements I {
  final value = 3;
}

void foo(Route a) {
    if (a is I) {
        print('${a.value}'); // Error: undefined_getter
    }
}
```

It also seem to be related to the fact that `Route` (from Flutter) has a generic type argument.

Is this the same underlying issue or should I file a new one?

**Comment by srawlins** [comment]
Sorry, @tp I think this is unrelated.

In your example, Route is unrelated to I, so you the programmer know that the type of `a`, inside your `if` body, is `Route & I`, but there is no concrete type that represents that intersection, so no type promotion is performed. I think https://github.com/dart-lang/sdk/issues/35314 is related to your request.

**Comment by jamesderlin** [comment]
I'm not sure if it's related, but a similar case:

```dart
class Foo {
  final int value = 42;
}

class Base<T> {}

class DerivedInt extends Base<int> {
  Foo get foo => Foo();
}

class Generic<T> {
  void f(Base<T> base) {
    if (base is DerivedInt) {
      print((base as DerivedInt).foo.value); // Works

      // The getter 'foo' isn't defined for the class 'Base<T>'.
      print(base.foo.value);
    }
  }
}

void main() {
  var derivedInt = DerivedInt();
  var instance = Generic<int>();
  instance.f(derivedInt);
}
```
In this case, the `is` check is fully qualified.  Why isn't the local variable `base` being promoted?

**Comment by eernstg** [comment]
You can only promote to a subtype, and `DerivedInt` is not a subtype of `Base<T>`. For instance, `T` could be `double` or `Widget`. Given that you're never using `T` you could use `void f(Base<Object?> base)`, and `DerivedInt <: Base<Object?>` does hold.

**Comment by jamesderlin** [comment]
@eernstg Thanks for the explanation.  I guess I have the wrong mental model for how promotion works.  I had been presuming that a `localVariable is T` check would be syntactic sugar for an automatic `localVariable as T` cast.  That is:
```dart
if (localVariable is T) {
  localVariable.f();
  localVariable.g();
}
```
would be equivalent to:
```dart
if (localVariable is T) {
  var temporary = localVariable as T;
  temporary.f();
  temporary.g();
}
```
So that raises a question: why shouldn't it be that way?  It seems to me that it would be easier to understand and would work the way more people expect and want.

The "only promote to a subtype" rule also seems weird and arbitrary since I could just upcast to `Object` first:
```dart
void f<T>(Base<T> base) {
  Object object = base;
  if (object is DerivedInt) {
    print(object.foo.value);
  }
}
```

**Comment by eernstg** [comment]
Promotion has been around in Dart at least since the language specification was added to the SDK repository in 2013, and it is required already in that 2013 version that the promotion changes the type of a local variable from a type `S` to a type `T` such that `T <: S`. This means that it would be a reversal of a very old decision to start promoting variables to anything other than a subtype of their previous type.

Promotion was made a lot more powerful in the version of Dart that supports null safety, but it still includes the requirement that the resulting type in a promotion is subtype related to the previous type. There is a list of 'types of interest' associated with the possible control flows up to the current location. We can promote to a subtype of the most recent one, and we can also demote to one of the earlier ones. Details [here](https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md).

However, even though it would be a serious breaking change to promote to an unrelated type, it could still be a good design choice. So let's consider that design choice, as if we could just do it without worrying about existing software.

The rationale we've used to justify restrictions on promotions is that they make the code less readable: Anyone who is reading source code using local variables (that is, just about any function/method body which is more than a few lines of code) will need to take the type of each local variable into account, in order to understand why it was created, and how to use it correctly.

Consider a variable `v` with declared type `T`. As long as promotions will promote `v` from `T` to some type `T1 <: T`, then to `T2 <: T1` and so on, the available members on `v` will be the ones on `T` plus a number of increments (some extras due to `T1`, some more due to `T2`, and so on). We may demote `v` by actions like `v = T2();` along the way, but we are still only using `v` with a type which is navigating the subtype graph under `T`. In particular, if we do `v.foo()` where `foo` is a member of `T` then we can rely on the assumption that this method call "will do the kind of work that `T.foo` does".

If we allow the type of `v` to change to something completely unrelated then we could promote `v` to a type `S` such that `S.foo` exists, but it has no relationship to `T.foo`. So we can completely misinterpret the action taken at `v.foo()`, and we'll have to scan the control flow backwards (for instance, we may need to find all the if-statements encountered up to this point, and various other constructs) in order to detect that the promotion to `S` occurred.

In other words, this rationale is all about being able to read a small snippet of code in a function body and understand what it does, rather than having to scan the whole function body up to that point and memorize several completely different interfaces that the variable might have.

Of course, there have been many, many other discussions about the design of variable promotion in Dart, but this one seems particularly important to me.

**Comment by cedvdb** [comment]
Might wanna move this to lang, I did not find this issue then [posted to stackoverflow](https://stackoverflow.com/questions/55093207/dart-type-checking-with-generics).


In some cases it's worse for readability 


```dart
   if( state is ReadStarted<Location, List<Restaurant>>) {
      
   }

```

vs


```dart
   if( state is ReadStarted) {

   }
```


I was expecting the property to be dynamic at least. But what would be better is if it was inferred from what's on the left side of the `is`

That is:

```dart
fn(ReadState<void, Location> readState) {
  if( state is ReadStarted) {
    // state is ReadState<void, Location> 
  }
```

If you want dynamic you have to explicitely set it.
